File: shared memory/my_ipc_4.cpp
// // //Use the shared memory and semaphores for the implementation of the logical ring-buffer
// // //(that can store up to 10 data items) and the synchronization.


#include<semaphore.h>
#include<fstream>
#include<iostream>
#include<thread>


using namespace std;

#define BUFFER_SIZE 10

int buffer[BUFFER_SIZE];

int in = 0; // points to next free element
int out = 0; // points to the first full buffer element
int count = 0; // counts the number of data items in the buffer
int producedCount = 0;

sem_t semEmpty;
sem_t semFull;

void producer_process() {
    ofstream generatedFile("producer_data_ipc_4.txt");
    while (producedCount < 100) {
        int x = (rand() % 100)+1;
        sem_wait(&semEmpty);
        buffer[in] = x;
        in = (in + 1) % BUFFER_SIZE; // Wrap-around behavior for the in pointer
        count++;
        producedCount++;
        sem_post(&semFull);
        //producedCount tells how many integers are generated.
        cout << producedCount << " - generated " << x << endl;
        generatedFile << x << endl; 
    }
    generatedFile.close();
}

void consumer_process() {
    ofstream consumedFile("consumed_data_ipc_4.txt");
    while (1) {
        sem_wait(&semFull);
        int y = buffer[out];
        out = (out + 1) % BUFFER_SIZE; // Wrap-around behavior for the out pointer
        count--;
        sem_post(&semEmpty);
        cout << "consumed " << y << endl;
        consumedFile << y << endl;
    }
    consumedFile.close();
}

int main() {
    srand(time(NULL));
    sem_init(&semEmpty, 0, BUFFER_SIZE - 1);
    sem_init(&semFull, 0, 0);

    thread producer(producer_process);
    thread consumer(consumer_process);

    producer.join();
    consumer.join();

    sem_destroy(&semEmpty);
    sem_destroy(&semFull);

    return 0;
}

File: shared memory/my_ipc_4.java
import java.util.concurrent.Semaphore;

class my_ipc_4 {
    private static final int BUFFER_SIZE = 10;
    private int[] buffer = new int[BUFFER_SIZE];
    private int in = 0; // points to next free element
    private int out = 0; // points to the first full buffer element
    private int count = 0; // counts the number of data items in the buffer
    private int totalGenerated = 0;

    private Semaphore semEmpty = new Semaphore(BUFFER_SIZE);
    private Semaphore semFull = new Semaphore(0);
    private Semaphore mutex = new Semaphore(1);

    public void producerProcess() {
        while (totalGenerated <= 100) {
            int x = (int) (Math.random() * 100);
            try {
                semEmpty.acquire();
                mutex.acquire();

                buffer[in] = x;
                in = (in + 1) % BUFFER_SIZE;
                count++;

                mutex.release();
                semFull.release();

                totalGenerated++;
                System.out.println("Generated: " + x);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    public void consumerProcess() {
        while (totalGenerated <= 100) {
            while (count == 0) {
                // Buffer is empty, so do nothing and wait
            }

            try {
                semFull.acquire();
                mutex.acquire();

                int y = buffer[out];
                out = (out + 1) % BUFFER_SIZE;
                count--;

                mutex.release();
                semEmpty.release();

                System.out.println("Consumed: " + y);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    public static void main(String[] args) {
        my_ipc_4 obj = new my_ipc_4();

        Thread producerThread = new Thread(obj::producerProcess);
        Thread consumerThread = new Thread(obj::consumerProcess);

        producerThread.start();
        consumerThread.start();

        try {
            producerThread.join();
            consumerThread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}
